eventbus.go

```diff
@@ -262,114 +180,105 @@ func (b *EventBus) AddHandler(ctx context.Context, m eh.EventMatcher, h eh.Event
+
+	if err := b.consumeConn.Close(); err != nil {
+		return fmt.Errorf(errMessage, err)
 	}
 
+	b.logger.logInfo("eventbus gracefully closed")
+
 	return nil
 }
 
+// Errors implements the Errors method of the eventhorizon.EventBus interface.
+func (b *EventBus) Errors() <-chan error {
+	return b.errCh
+}
+
 // Handles all events coming in on the channel.
 func (b *EventBus) handle(
-	consumer *rabbitmq.Consumer,
+	consumer *clarimq.Consumer,
 ) {
 	defer b.wg.Done()
 
-	<-b.cctx.Done()
+	<-b.ctx.Done()
 
+	b.consumerMu.Lock()
 	consumer.Close()
+	b.consumerMu.Unlock()
 }
 
-// ErrDiscardEvent is used to drop an event manually.
-var ErrDiscardEvent = errors.New("drop event from eventbus")
-
 func (b *EventBus) handler(
 	ctx context.Context,
-	m eh.EventMatcher,
-	h eh.EventHandler,
-	dlxName, dlQueueName string,
-) func(d rabbitmq.Delivery) rabbitmq.Action {
-	return func(msg rabbitmq.Delivery) rabbitmq.Action {
-		event, ctx, err := b.codec.UnmarshalEvent(ctx, msg.Body)
+	matcher eh.EventMatcher,
+	handler eh.EventHandler,
+) func(d *clarimq.Delivery) clarimq.Action {
+	return func(msg *clarimq.Delivery) clarimq.Action {
+		event, ctx, err := b.eventCodec.UnmarshalEvent(ctx, msg.Body)
 		if err != nil {
-			b.sendErrToErrChannel(ctx, err, h, event)
+			b.sendErrToErrChannel(ctx, err, handler, event)
 
-			if b.useRetry {
-				action, err := b.deadLetterMessage(&msg, dlxName, dlQueueName)
-				if err != nil {
-					b.sendErrToErrChannel(ctx, err, h, event)
-				}
-
-				return action
-			}
-
-			return rabbitmq.NackDiscard
+			return clarimq.NackDiscard
 		}
 
 		// Ignore non-matching events.
-		if !m.Match(event) {
-			return rabbitmq.Ack
-		}
-
-		if b.useRetry {
-			retryCount, ok := msg.Headers[retryCountHeader].(int32)
-			if !ok {
-				retryCount = 0
-			}
-
-			ctx = NewContextWithNumRetries(ctx, int64(retryCount))
+		if !matcher.Match(event) {
+			return clarimq.Ack
 		}
 
 		// Handle the event if it did match.
-		if err := h.HandleEvent(ctx, event); err != nil {
-			b.sendErrToErrChannel(ctx, err, h, event)
+		if err := handler.HandleEvent(
+			ehtracygo.NewContext(ctx, msg.CorrelationId),
+			event,
+		); err != nil {
+			b.sendErrToErrChannel(ctx, err, handler, event)
 
-			// discard event manually
-			if errors.Is(err, ErrDiscardEvent) {
-				return rabbitmq.NackDiscard
-			}
-
-			if b.useRetry {
-				action, err := b.deadLetterMessage(&msg, dlxName, dlQueueName)
-				if err != nil {
-					b.sendErrToErrChannel(ctx, err, h, event)
-				}
+			return clarimq.NackDiscard
+		}
 
-				return action
-			}
+		return clarimq.Ack
+	}
+}
 
-			return rabbitmq.NackDiscard
-		}
+func (b *EventBus) returnHandler(rtn clarimq.Return) {
+	event, ctx, err := b.eventCodec.UnmarshalEvent(b.ctx, rtn.Body)
+	if err != nil {
+		b.logger.logDebug("return handler: failed to unmarshal event", "error", err)
 
-		return rabbitmq.Ack
+		return
 	}
+
+	b.logger.logError("return handler: event could not be published",
+		"tenant", namespace.FromContext(ctx),
```

Comment:
```suggestion
		"eh-namespace", namespace.FromContext(ctx),
```

tenant is maybe too specific for our use case 😉